addons/sale_timesheet/models/sale_order.py

```diff
@@ -72,6 +73,32 @@ def _compute_field_value(self, field):
                 # We want to display only one time the warning for each SOL
                 upsellable_lines.write({'has_displayed_warning_upsell': True})
 
+    def _compute_show_hours_recorded_button(self):
+        if not self.env.user.has_group('hr_timesheet.group_hr_timesheet_user'):
+            for order in self:
+                order.show_hours_recorded_button = False
+            return
+
+        orders_to_check = self.filtered(lambda l: l.state not in ['draft', 'sent'])
+        list_order = self.env['sale.order.line']._read_group([('order_id', 'in', orders_to_check.ids)], ['product_id:array_agg(product_id)'], ['order_id'], lazy=False)
+        dict_order = {}
+        sol_product_ids = set()
+        for order in list_order:
+            product_set = set(order['product_id'])
+            dict_order[order['order_id'][0]] = product_set
+            sol_product_ids.update(product_set)
+        domain = [('id', 'in', list(sol_product_ids)), ('type', '=', 'service'), '|', ('service_type', 'not in', ['milestones', 'manual']), ('invoice_policy', '!=', 'delivery')]
+        products_service_prepaid_or_timesheet = self.env['product.product'].search(domain)
+        product_ids = set(products_service_prepaid_or_timesheet.ids)
+        for order in self:
+            order.show_hours_recorded_button = order._has_data()
+            if order.show_hours_recorded_button:
+                order.show_hours_recorded_button = order in orders_to_check and not product_ids.isdisjoint(dict_order[order.id])
```

Comment:
```suggestion
            order.show_hours_recorded_button = order.project_count\
                and order in orders_to_check\
                and not product_ids.isdisjoint(dict_order[order.id])
```
Unless I'm missing something, I think this is equivalent to your code?

Reply:
I added the _has_data function, because i overwrite it in helpdesk_sale_timesheet. There the condition to set the field change. And i thought it was better to have that small extra function here in order to set directly the correct value on the field, instead of having to make an extra function in helpdesk_sale_timesheet to update the field if needed.